Paper: Unsupervised Spatiotemporal Representation of Cilia Video Using A Modular Generative Pipeline

Micky774 opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

Micky774 committed:
Initial commit

Micky774 committed:
More recent version, ported from overleaf

Micky774 committed:
Ready for initial submission

deniederhut commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,702 @@
+captured from a top-down perspective without possibility of dilation, we limit our computation to curl 
+and deformation, similar to Quinn 2011 :cite:`quinn_novel_2011`. 
+	
+.. figure:: of_Ex.png
```
Hi! It looks like the current built is crashing because the filename has a lowercase e, like `of_ex.png`. Would you mind updating this?

Micky774 replied:
Updated!

deniederhut replied:
Thanks!

Micky774 committed:
Fixed file name typo 'of_Ex'

Micky774 committed:
Updated dyn figure settings

Micky774 committed:
Fixed broken figure reference (updated fig label)

Micky774 committed:
Moved appearance pipeline figure

Micky774 committed:
Moved section header / trimmed

lizchenx committed:
Update zain.rst

line 346, "referring you to [?] for further reading", is this [?] supposed to be that way?

lizchenx committed:
Update mybib.bib

added the VAE ref [kingma2013auto]

lizchenx committed:
Update zain.rst

inserted citation 'kingma2013auto' on line 289

saketkc commented:
Hi @Micky774, I will be reviewing your paper! I will leave my comments below over the weekend. Thanks!

snikumbh commented:
Hi @Micky774, I look forward to reviewing your paper. Comments coming very soon. Thanks :)

Micky774 committed:
Added/Fixed citations

Micky774 committed:
Added more citations

Micky774 commented:
Looking forward to working with you all!

deniederhut commented:
Hey @Micky774 ! Thanks to some awesome behind the scenes work by @stargaser, we now have experimental support for publishing papers with ORCIDs. If you have one and would like to add it to your paper, it's as simple as adding an `:orcid:` tag under your name in the paper header. [You can see an example of how to do this in the `00_vanderwalt` example paper](https://raw.githubusercontent.com/scipy-conference/scipy_proceedings/2020/papers/00_vanderwalt/00_vanderwalt.rst). If you don't have one, no worries! You can still publish your paper as is.

deniederhut commented:
Hi @saketkc and @snikumbh ! How are your reviews coming along?

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+generic dynamics models fail to classify and cluster this type of motion accurately or meaningfully, 
+and are insufficient for generating a semantically potent representation. We thus apply a novel 
+machine learning approach to create an underlying representation which then can be used for downstream 
+tasks such as classification and clustering, and whatever else experts may deem necessary. Furthermore, 
```
Consider changing "whatever else" => "any other tasks"

Micky774 replied:
Changed, thanks

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+Therefore, we aim to develop an unsupervised, computational approach to analyze ciliary motion, 
+developing a quantitative "library" of well-defined, clinically relevant ciliary motion phenotypes. 
+Clustering and classification are established problems in machine learning. However, their 
+applications to ciliary waveform analysis are novel: cilia exhibit subtle, rotational, non-linear 
```
It is not clear if this statement implies that application to ciliarly waveform analysis are novel because of their subtle/rotational/non-linear motion or just in general. The authors do point out Lu 2018 and Quinn 2015 (both from the same group). I believe the reader can benefit from clarifying that clustering and classification in this setting is more nuanced because of the particular properties attached with it (rotational/non-linear motion for example). This is also highlighted in the statement that appears later "While attempts have been made at addressing this problem,"

Micky774 replied:
Updated that portion to clarify the difficulty in immediately applying those techniques, rather than saying the application itself is novel. We really want to talk about the fact that you can't quite obviously "plug 'n chug" a good solution by applying generic clustering algorithms, so I think that's better conveyed now.

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+and are insufficient for generating a semantically potent representation. We thus apply a novel 
+machine learning approach to create an underlying representation which then can be used for downstream 
+tasks such as classification and clustering, and whatever else experts may deem necessary. Furthermore, 
+we avoid using labeled data in order to free the model from systematic assumptions naturally imposed by 
```
It would be useful to briefly describe what "labeled data" is in this context. Disease state, abnormal mobility etc.

Micky774 replied:
Attempted a fix. Let me know if you think it's worth diving into more detail regarding labeled data. Thanks

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+
+We aim to build upon these studies by developing a fully unsupervised approach to characterizing 
+ciliary motion phenotypes. This pipeline is advantageous in that it does not need hand-drawn 
+ROI maps nor a labelled dataset as training data. While clinicians acknowledge the existence of 
```
"labelled" => "labeled"

Micky774 replied:
Fixed

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+that the study is ostensibly attempting to remove. 
+
+The model proposed by Quinn 2015 was improved upon by Lu 2018, the latter attempt using stacked 
+Fully Convolutional DenseNets and LSTM networks :cite:`lu_stacked_2018`.  Densely Connected 
```
Consider citing LSTM publication as well (DenseNet already cited).

saketkc replied:
Ideally, Lu 2018 should be followed by the citation rather than at the end of statement.

Micky774 replied:
Updated. Moved DenseNet initial citation as well to maintain consistency.

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+of ciliary motion. Unsupervised representation learning enables a model to learn families of 
+complex ciliary motion phenotypes beyond the normal-abnormal binary.
+
+Methods
```
The Methods section as it currently stands is just a paragraph long. The Results section on the other hand is long. But it also appears that part of the Results section is actually Methods. The authors should consider reorganizing the text in these two sections. The focus of the Methods section should be to describe the details of the architecture, its working and how parameters/hyperparameters were chosen. The Results section should focus on the evaluation of this model on their dataset.

snoiarao replied:
Restructured such that Methods contains most of prior Results section; let us know if this requires further structural changes!

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+Methods
+-------
+
+Our proposed model is divided into three modules: preprocessing, appearance, and dynamics. 
```
The authors might consider adding a flowchart that explains the three modules briefly as Figure 1.

Micky774 replied:
Working on that, will update when added.

Micky774 replied:
Added

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+Each expanded video is fed frame-by-frame to the appearance module which learns a compressed spatial 
+representation for images of cilia. After the appearance module learns a latent representation of 
+the spatial aspects of cilia in that frame, videos are embedded as sequences of points in the compressed 
+latent space. The dynamics module employs another VAE to learn a representation from this compressed 
```
The way the dynamics sections is described here, it seems it is another appearance section as only the VAE part is described but the dynamics part of this module is missing.

snikumbh replied:
This is the first usage of the acronym VAE, please define it here.

Micky774 replied:
VAE definition has been updated. Added some more explanation regarding the dynamics module. Let me know if you think it suffices, or if it needs more. Thanks!

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+motion of cilia. Through this construction, we factor the representation of cilia into disentangled 
+spatial and temporal components.
+
+Results
```
As mentioned in an earlier comment, the authors should consider reorganizing the text to separate the methods from results. A lot of the content in the next few sections belongs to Methods in my opinion.

snoiarao replied:
Restructured!

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+=============
+
+The preprocessing module primarily functions to
+- generate segmentation masks that distinguish spatial regions containing cilia from background noise and 
```
- before generate seems unnecessary.

Micky774 replied:
Changed

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+is treated as a time series using convolutional long short-term memory (Conv-LSTM) networks, a 
+specific type of recurrent neural network (RNN), to model the long-term temporal dependencies in the data. 
+
+We aim to build upon these studies by developing a fully unsupervised approach to characterizing 
```
I am a bit unsure about claiming this to be fully unsupervised approach given the segmentation task does use ground truth masks and microscopy data for training where the ground truth masks are generated manually. How about calling it a "minimally supervised" or semi-supervised approach. This would also mean that the title of the paper be changed accordingly. I am sure the authors have given it a thought so it might make sense to explain the rationale for calling it fully unsupervised. This could be done, for example, by justifying that the trained segmentation module is generalizable.

Micky774 replied:
Added clarification in the preprocessing methodology section addressing the supervised nature of the current segmentation, as well as changed the title to "Towards ..." to more clearly communicate the lack of a fully unsupervised network. I think it would be okay to leave the section you highlighted as is since it specifies that we "aim to..." develop a fully unsupervised approach. Let me know your thoughts, and whether you think that anything further needs to be done. Thanks!

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+of cilia move synchronously, slowly, and within a set spatial region near cell boundaries. Additionally, 
+our imaging modality allowed for consistent object brightness throughout sequences of frames. As such, 
+we explored optical flow solutions that focus on brightness constancy, small motion, and spatial coherence 
+systems of equations. Among Farneback, Horn-Schunck, and Lucas-Kanade optical flow computation algorithms, 
```
Farneback, Horn-Schunck, and Lucas-Kanade  are all missing citations (HS81 citation appears later, but should be ideally moved here).

snoiarao replied:
Fixed

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+captured from a top-down perspective without possibility of dilation, we limit our computation to curl 
+and deformation, similar to Quinn 2011 :cite:`quinn_novel_2011`. 
+	
+.. figure:: of_ex.png
```
Consider improving the image quality (higher dpi, bolder/bigger fonts).

Micky774 replied:
Improved size of the specific image, as well as changing it to a vertical stack to better fit the columns. Updating font size on a couple other images as well.

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+    
+	Raw imagery and corresponding optical flow visualization :label:`of`
+
+:ref:`of` shows an example of healthy cilia and its mid-cycle optical flow where vector magnitude 
```
At the least this statement should not begin with a citation (Consider adding 'Figure'). Also this is probably the first "Result" in this section with the rest being more suitably part of the "Methods" section.

Micky774 replied:
Added 'Figure' before citation. Keeping conversation open pending restructuring of methods/results

snoiarao replied:
Moved paragraph and figure to results section

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+pipeline. Our current choice of generative model is a variational autoencoder (VAE) 
+:cite:`introVAE`, an architecture that generates a low-dimensional representation of 
+the data, parameterized as a probability distribution. A VAE can be considered a modified autoencoder 
+(AE). A general autoencoder (AE) attempts to learn a low-dimensional representation of the data by 
```
AE abbreviation doesn't need to be defined twice.

Micky774 replied:
Fixed

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+	E_\theta:\mathbb{R}^n\rightarrow\mathbb{R}^l,\quad D_\theta:\mathbb{R}^l\rightarrow\mathbb{R}^n
+
+The subscript :math:`\theta` denotes that these functions are constructed as neural networks parameterized 
+by learanble weights :math:`\theta`.The encoder is tasked with taking the original data input and 
```
Add space before "The encoder"

Micky774 replied:
Fixed

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+covariance be a diagonal matrix, allowing us to write :math:`\Sigma=\operatorname{diag}(\sigma)` 
+for some :math:`\sigma\in\mathbb{R}^l`. While the decision to model neighborhoods via distributions 
+deserves its own discussion and justification, it falls outside the scope of this paper and thus 
+we omit the technical details while referring you to :cite:`doersch2016tutorial` for further reading. Instead, we provide 
```
"you to" => "the readers to".

Micky774 replied:
Fixed

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+
+.. math::
+
+	L_\theta(x)=\|x-D_\theta(E_\theta(x))\|^2
```
End equation with a '.' (applies elsewhere too where it is end of a statement).

Micky774 replied:
Fixed

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+
+.. math::
+
+	L_\theta(x)=\|x-D_\theta(z)\|^2
```
End with a '.'.

Micky774 replied:
Fixed

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+in that there's no guarantee that the decoder doesn't degenerate to setting :math:`\sigma` arbitrarily 
+close to zero, resulting in a slightly more complex AE. Thus we assert that if one were to sample from 
+some predetermined *prior distribution* on the latent space, written as :math:`p(z)`, then the sampled 
+point should be able to be reasonably decoded as a point in the starting data space. To break that down, 
```
Shorten "should be able to be " => "can be". Also, a lot of content in this section should be moved to the "Methods" seciton.

Micky774 replied:
Updated wording.

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+
+.. math::
+
+	L_\theta(x)=\|x-D_\theta(z)\|^2+\operatorname{KL}(q_\theta(z|x) \,\|\, p(z))
```
End with a .

Micky774 replied:
Fixed

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+back into video frames through the decoder of the appearance module :math:`D_\text{app}`.
+ 
+Since the encoder and the decoder of the dynamics module need to process sequences of vectors, they 
+are modeled using a Gated Recurrent Unit (GRU:cite:`Cho2014LearningPR`) and a Long Short-Term Memory 
```
The cite here is not being parsed properly. LSTM can be cited early on (See my previous comments). The LSTM abbreviation also should appear earlier.

Micky774 replied:
Updated and fixed, in tandem with previous comment.

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+Implementation Notes
+====================
+
+We train our FCDN-103 model, written in PyTorch, with an Adam optimizer and cross-entropy 
```
Please cite Pytorch. This applies to all the software libraries you are using. Please.

Micky774 replied:
Fixed

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+smoothness assumption, we observe that this is mitigated by the coarse-to-fine estimation and 
+hyperparameter tuning. Additionally, we find that this estimation is more computationally and 
+time efficient than its contemporaries. Curl and deformation fields are extracted from the 
+generated optical flow fields using SciPy's signal and ndimage packages. 
```
Please cite Scipy.

Micky774 replied:
Fixed

saketkc commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+
+During optimization of the appearance module, we observe that cilia do not tend to exhibit a 
+large degree of spatial differences over time, thus rather than processing every frame of the dataset, 
+we used the NumPy library to efficiently sample a fixed number of frames from each video. 
```
Cite NumPy.

Micky774 replied:
Fixed

saketkc reviewed (changes requested):
Hi @Micky774, thanks for submitting the full paper. I learned a lot in the process. Most of my comments appear inline. But a general comment: the paper can be improved to make it an easier read. In particular, the text needs to be organized so that the methods and results are clearly distinguished. At places the paper goes into too many details which I believe can all be combined into a smaller section. The general idea should be to give an overview of the methods first before showing the results and then elaborating on the underlying ideas in the methods in a new section. 

Happy to answer any questions you or other authors might have after reading my comments. Just one last general comment, please consider citing all software you are using. I was a bit discouraged to see particularly that Scipy's citation was missing.

Thanks,
Saket

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+consequences can include dysfunction of macroscopic structures within the lungs, kidneys, 
+brain, and other organs. Characterizing ciliary motion phenotypes as healthy or diseased 
+is an essential step towards diagnosing and differentiating ciliopathies. We propose a 
+modular generative pipeline for the analysis of cilia video data so that expert labor may 
+be supplemented or replaced for this task. Our proposed model is divided into three modules: 
```
"replacing" the expert can be debatable. Given the results and the conclusion section text, once the complete pipeline is implemented and thoroughly tested to perform satisfactorily, this could be discussed.

Micky774 replied:
Agreed. Initially the statement was meant to convey the idea of shifting the _focus_ of expert labor rather than necessarily replacing it, though "supplemented" sufficiently conveys that I think. Removed "replaced"

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+tasks such as classification and clustering, and whatever else experts may deem necessary. Furthermore, 
+we avoid using labeled data in order to free the model from systematic assumptions naturally imposed by 
+labels |---| the choice of labels themselves can inadvertently limit the model by asserting that all data 
+*must* conform to those exact labels. An unsupervised model has the freedom to discover potentially 
+semantically meaningful patterns and phenotypes that fall outside current clinical thinking. Furthermore, 
```
change "... discover potentially semantically ..." to "... discover potential semantically meaningful patterns ..."

Micky774 replied:
Changed

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+Our approach is to create a pipeline that learns a low-dimensional representation of ciliary motion on 
+unlabeled data. The model we propose considers the spatial and temporal dimensions of ciliary motion 
+separately. The pipeline encodes each frame of the input video and then encodes the paths between frames in the latent space. 
+The low-dimensional latent space in this pipeline will have semantic significance, and thus interpolation 
+and clustering on the latent space should be meaningful for those studying ciliary motion and its connection to ciliopathies. 
```
I am not sure if I understood the role of "interpolation" here. Did you mean interpretation by any chance?

Micky774 replied:
Ah that's a vestige of an older version of this paper, good catch. Originally we were going to do more with interpolation on the latent space, but decided to focus instead on a VampPrior based approach which doesn't ensure meaningful interpolation. Updating.

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+dynamic texture, which are rhythmic motions of particles subjected to stochastic noise :cite:`doretto2003` 
+and include familiar patterns such as flickering flames, rippling water, and grass in the wind. 
+Each instance of dynamic texture contains a small amount of stochastic behavior altering an 
+otherwise consistent visual pattern. Likewise, ciliary motion can be considered a dynamic 
+texture as it consists of rhythmic behavior subject to stochastic noise that collectively 
+determine the beat pattern. The authors used autoregressive (AR) representations of optical 
```
The sentence beginning with "Likewise, ..." should also be phrased in the past tense because that is something the earlier work considered/did (if I understood correctly).

Micky774 replied:
Rewrote sentence to incorporate suggestion as well as (hopefully) clarify the role of the referenced authors.

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+ROI maps nor a labelled dataset as training data. While clinicians acknowledge the existence of 
+distinct ciliary waveform phenotypes beyond "normal" and "abnormal", experts lack standard 
+guidelines for qualitatively or quantitatively categorizing ciliary beat pattern. Additionally, 
+experts may not observe the level of quantitative detail required to associate complex motion 
+phenotypes with specific ciliopathies and genetic mutations :cite:`quinn2015automated`. 
```
I think you mean the inability to observe complex non-linear relationships in data in general?

Micky774 replied:
Sure, but that section was heavily contextualized as specifically looking at the case of ciliopathies, as the statement was also supported by the referenced material, so I don't know whether it'd be appropriate to generalize the statement. Do you think it's necessary or worthwhile?

snikumbh replied:
I see. Then, leaving that way is fine.

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+experts may not observe the level of quantitative detail required to associate complex motion 
+phenotypes with specific ciliopathies and genetic mutations :cite:`quinn2015automated`. 
+Thus, we shift away from a classification-style task (classifying abnormal versus normal ciliary 
+motion) to a representational learning task to generate meaningful, low dimensional representations 
```
consistently hyphenated --> low-dimensional

Micky774 replied:
Fixed

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+masks and extracting dense optical flow vector fields and pertinent differential quantities. 
+Segmentation masks are used to limit spatial representation learning to video regions containing 
+cilia, and optical flow fields are computed from consecutive frames as a compressed representation 
+of temporal behavior. The predicted segmentation masks and optical flow entities are concatenated 
+with the original video data as additional channels to each frame to form an augmented video. 
```
How does the quality of prediction of the segmentation masks affect the results downstream? Has this been studied yet?

Micky774 replied:
Nope. A very good question, which I've now included in the further research section (as well as a couple similar questions)

Micky774 replied:
Let me know whether or not this sufficiently resolves the conversation, thanks!

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+Because we are interested in modelling the spatio-temporal behavior of *only* cilia, segmentation masks, 
+which provide a direct mapping to pixels of interest within each frame, are critical within the appearance 
+module to limit representation learning to cilia localities and ignore background noise. Although 
+the end-to-end pipeline provides an unsupervised framework to represent and characterize complex and 
+dynamic ciliary motion phenotypes, this module utilizes *supervised* segmentation to produce initial 
+segmentation masks. Because we do not have ground-truth segmentation masks for every sample in our dataset, 
+a supervised network allows us to augment our set such that each raw video has a corresponding segmentation 
+mask to be used in subsequent modules. We draw upon prior supervised segmentation literature to implement 
```
Using the few ground truth segmentation masks to predict segmentation masks for all samples in the data for further use downstream, qualifies as a semi-supervised approach instead of unsupervised. While I agree that this is the case only for the preprocessing module, if the downstream performance is heavily dependent on this, the authors should consider calling their approach semi-supervised.

Micky774 replied:
Added section addressing this point, as well as changed the title to "Towards ..." to more clearly communicate the lack of a fully unsupervised network.

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+straightforward upsampling path requires multiplication of high-resolution feature maps, resulting in a 
+computationally intractable number of feature maps. To mitigate this "feature explosion" issue, 
+FCDenseNets upsample only the preceding dense block instead of upsampling all feature maps concatenated 
+in previous layers. We modify and train a FCDenseNet to generate usable segmentation masks as input to 
+the appearance module. Our architecture, shown in :ref:`dense`, consists of dense blocks, transition 
```
It is not clear which parts of FCDenseNet architecture are modified.

snoiarao replied:
Sorry, ambiguous wording. Fixed

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+be thought of as the projection of three dimensional motion on an image plane, relative to a 
+visual sensor such as a camera or microscope. As such, optical flow represents the apparent 
+motion of pixels within consecutive frames, relative to the visual sensor. To calculate pixel 
+displacement, optical flow algorithms are contingent on several assumptions. Brightness constancy 
+assumes that a pixel's apparent intensity does not change between consecutive frames, small 
+motion assumes that pixels are not drastically displaced between consecutive frames, spatial and 
+temporal coherence assumes that a pixel's neighbors likely exhibit similar motion over gradual time. 
```
Consider numbering the assumptions for more readability instead of a single long sentence.

Micky774 replied:
Changed

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+our imaging modality allowed for consistent object brightness throughout sequences of frames. As such, 
+we explored optical flow solutions that focus on brightness constancy, small motion, and spatial coherence 
+systems of equations. Among Farneback, Horn-Schunck, and Lucas-Kanade optical flow computation algorithms, 
+we incorporate a slightly modified Horn-Schunck algorithm that adequately captures synchronous ciliary 
+motion :cite:`horn_determining_1981`.
```
A curious reader would wonder, why are the other algorithms not suitable or why was the Horn-Shunck algorithm chosen?
What is the modification in the Horn-Schunck algorithm?

snoiarao replied:
There is a brief explanation in the 'Technical Discussion: Implementation Details' subsection, but it will be moved up once Methods section is restructured. Will update once moved

snoiarao replied:
Moved explanation of optical flow to methods section

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+information based representation of the input data. The size of the bottleneck is a 
+hyperparameter which determines how much of the data is compressed.
+
+With this task in mind, an AE can be considered as the composition of two constituent neural networks: 
+the encoder, and the decoder. Suppose that the starting dataset is a collection of n-dimensional points, 
```
This paragraph can be subsectioned as brief introduction to AEs and VAEs.

Micky774 replied:
Changed the section structure to hopefully organize content more clearly.

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+	L_\theta(x)=\|x-D_\theta(z)\|^2
+	
+Through this change, over the course of training we obtain a Monte Carlo estimation of the neighborhoods 
+around the embedded points, encouraging continuity in their decoding. This results is still incomplete 
+in that there's no guarantee that the decoder doesn't degenerate to setting :math:`\sigma` arbitrarily 
```
typo? "results" --> "result"

Micky774 replied:
Yup, fixed, thanks!

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+point should be able to be reasonably decoded as a point in the starting data space. To break that down, 
+this means that the portions of the latent space that our model should be best trained on should follow 
+the prior distribution. A common choice for prior, due to simplicity, is the unit-variance Gaussian 
+distribution. This is implemented by imposing a KL-Divergence loss between the posterior distributions 
```
First instance of KL-Divergence usage so define the acronym as "Kullback–Leibler".

Micky774 replied:
Fixed. Also updated KL-Divergence to KL Divergence to stay consistent with literature.

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+includes aspects such as the shape and distribution of cilia along cells, as well as 
+factors such as their length, orientation and overall shape. These spatial features can be 
+then used in downstream tasks such as phenotype discovery, by drawing a strong connection 
+between apparent patterns in this space as and semantically meaningful patterns in the 
```
"apparent patterns in this space" --> consider replacing "this" with the space you are referring to for more clarity.

Micky774 replied:
Changed

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+ 
+Since the encoder and the decoder of the dynamics module need to process sequences of vectors, they 
+are modeled using a Gated Recurrent Unit (GRU:cite:`Cho2014LearningPR`) and a Long Short-Term Memory 
+(LSTM:cite:`LSTM1999`) unit, respectively. They are types of recurrent neural networks (RNN) with 
```
The acronym RNN is already defined above, so can be omitted here.

Micky774 replied:
Fixed

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+would encompass information about the motion patterns in the observed video frames :math:`z_{1..n}`, 
+and would then constitute a latent vector in the dynamics semantic space :math:`w=s^n`.
+
+The dynamics decoder LSTM :math:`D_\text{dyn}` starts from from a latent dynamics vector as its hidden 
```
typo: from appears twice

Micky774 replied:
Fixed

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+
+	z\sim q_\theta(z|x) \iff z=\mu+\Sigma\epsilon\text{, where } \epsilon\sim\mathcal{N}(0,\mathcal{I}_l)
+	
+to ensure that we may complete backpropogation, since :math:`\mu,\sigma` are dependent on weights 
```
typo: backprop_o_gation --> backprop_a_gation

Micky774 replied:
Fixed. One day I will get that word right on the first try. Thanks!

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+sequence into a dynamics latent vector :math:`w`, multiple sequences of :math:`k+1` frames are sampled 
+uniformly from the same training video, where each generated sequence is set to extrapolate from its 
+first frame, and the same dynamics vector :math:`w`. The :math:`L2` loss between the extrapolated frames 
+and the remaining :math:`k` frames in each sequence is minimized with back-propagation.
```
previous usage of backpropagation appeared w/o a hyphen

Micky774 replied:
Fixed

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+standard parameters such as batch size, learning rate, and regularization parameters such as 
+learning rate decay, weight decay, and dropout. We observe optimal performance after 50 epochs, 
+14 patches per batch, learning rate of 0.0001, learning rate decay of 0.0, and weight decay of 0.0001. 
+This model achieves 33.06% average testing IOU, and 53.26% precision. Figure :ref:`masks` 
+shows two examples of 128 x 128 test patches with their corresponding ground truth mask (middle) 
+and FCDN-103 generated mask (right); the predicted masks cover more concise areas of cilia 
```
This information is suitable for the results section.

snoiarao replied:
moved to results

snikumbh commented on papers/zain/zain.rst:
```diff
@@ -0,0 +1,703 @@
+into the realm of collaborative, crowd-driven data acquisition and processing in the form of a user-friendly web tool.
+
+We also encourage the expansion and application of this framework to various other problem contexts. 
+The modular approach to its design ensures portability and adaptability to all sorts of projects. 
```
"all sorts of projects" --> other projects.

Micky774 replied:
Changed

snikumbh reviewed (commented):
Hi @Micky774 and co-authors,

Congratulations on the full paper submission. I enjoyed reading the paper. Here are some suggestions towards improving the paper further.

I found that the general idea of the paper -- the problem and the proposed solution -- is put forth well. However, some important aspects of the paper need work. 
1. The paper is a good read in parts and details. Meaning that, each paragraph or portion of text explains the technicalities well individually. Proper sectioning/structuring would be very helpful to clearly convey the overall goal and status of the work. For instance, the relative work section could end with a subsection clearly delineating the contributions of this work.
2. A proper, complete methods section giving details of all modules at once in respective subsections can help the reader. Beginning this section with a flow diagram pictorially explaining the modules would be very beneficial. 
3. All figure legends can be expanded with some more text/details of what the user should observe/note in the figure. Also, one could use this space to explain all acronyms appearing in the figure.
4. After the results section, conclude with a discussion on the current status of project implementation, attained results, future work. Some details from technical discussion section could be moved to the results section (see individual comments).

Further comments inline.
Please feel free to ask if something is not clear.

Micky774 commented:
Thank you both for your feedback! We'll be taking a detailed look at it and responding soon!

Micky774 committed:
Merge branch '2020' into 2020

Micky774 committed:
Merge pull request #1 from lizchenx/2020

Minor changes

Micky774 committed:
Initial round of review/editing

snoiarao committed:
comment L210

snoiarao committed:
opflow citations

snoiarao committed:
farneback citation

snoiarao committed:
opflow citations

snoiarao committed:
of_ex vertical stack

Micky774 commented:
Also, does anyone know how to add/change the corresponding author? I can't quite figure it out on the rst side of things

Micky774 committed:
Added figure, along with minor edits

snoiarao committed:
structural changes

snoiarao committed:
opflow considerations

snoiarao committed:
wording

snoiarao committed:
hs citation

lizchenx committed:
Merge pull request #1 from Micky774/2020

update from the original

snoiarao committed:
OF fig to results

lizchenx committed:
remove "introduction" under appearance module

remove "introduction" under appearance module, add title "discussion"

Micky774 committed:
Merge pull request #2 from lizchenx/2020

remove 'introduction' under generative module

Micky774 committed:
Expanded methods intro and relabeled sections

Micky774 committed:
Trimmed/Resized figure, added clarifying statements

Micky774 committed:
Updated intial VAE definition

Micky774 commented:
@deniederhut It seems that our paper is failing to build on the scipy server, though it builds locally perfectly fine. Any idea what the issue may be?

Micky774 committed:
Updated graphics

nssafir committed:
Update zain.rst intro

Responded to comment that classification and clustering are not necessarily novel approaches to ciliary motion problem.

Micky774 committed:
Clarified wording in intro

Micky774 committed:
Label change and clarification

deniederhut commented:
👀

Looks like latex is tripping on this author name: `Çaglar G{\"u}lçehre` Here's the actual error:

```
! Package inputenc Error: Unicode char �G (U+4A)
(inputenc)                not set up for use with LaTeX.

See the inputenc package documentation for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.24 ...[CvM�G{\etalchar{+}}14]{Cho2014LearningPR}
```

I can look into this later. Until we get this resolved, you might want to swap out the Ç for C.

snikumbh commented:
Simply use `{\c{C}}aglar G{\"{u}}l{\c{c}}ehre` from the bibtex entry at DBLP:

> @article{DBLP:journals/corr/ChoMGBSB14,
  author    = {Kyunghyun Cho and
               Bart van Merrienboer and
               {\c{C}}aglar G{\"{u}}l{\c{c}}ehre and
               Fethi Bougares and
               Holger Schwenk and
               Yoshua Bengio},
  title     = {Learning Phrase Representations using {RNN} Encoder-Decoder for Statistical
               Machine Translation},
  journal   = {CoRR},
  volume    = {abs/1406.1078},
  year      = {2014},
  url       = {http://arxiv.org/abs/1406.1078},
  archivePrefix = {arXiv},
  eprint    = {1406.1078},
  timestamp = {Mon, 13 Aug 2018 16:46:44 +0200},
  biburl    = {https://dblp.org/rec/journals/corr/ChoMGBSB14.bib},
  bibsource = {dblp computer science bibliography, https://dblp.org}
}

Micky774 committed:
Updated bibfile

Micky774 commented:
Thank you both very much! All seems good now.

Micky774 committed:
Updated figures to pdf

Micky774 committed:
Added missing ResNet citation

Micky774 committed:
Updated corresponding author

Micky774 commented:
@saketkc @snikumbh I believe all your comments have either been fully adopted/resolved or at least addressed and are being kept open pending your input. Thank you for all your well thought out and helpful feedback -- it's definitely made a huge difference in the quality of the paper. Let me know if there's anything else you'd like us to change!

deniederhut commented:
Hi @snikumbh and @saketkc ! Are you happy with the updates made to the paper?

saketkc commented:
I am happy with the changes @Micky774 and @deniederhut. The authors have addressed all my comments (including the title!) and I would like to applaud them for their efforts! I agree, the content and quality of paper has tremendously improved.

 I just have a couple of final recommendations:
 - I would suggest the authors consider open sourcing their work. A link to a repository would add a lot of value to the paper. (I understand this might not be possible)
- The figures seem to have been generated using matplotlib, in which case matplotlib needs a [citation](https://www.scipy.org/citing.html#matplotlib)

Congrats @Micky774, @snoiarao, @nssafir, @lizchenx and all the other authors - Great work!

snikumbh commented:
Happy with all the updates! 
Congratulations to the authors/contributors!

Micky774 committed:
Added matplotlib citation and current repo

Micky774 commented:
> I just have a couple of final recommendations:
> 
> * I would suggest the authors consider open sourcing their work. A link to a repository would add a lot of value to the paper. (I understand this might not be possible)
> * The figures seem to have been generated using matplotlib, in which case matplotlib needs a [citation](https://www.scipy.org/citing.html#matplotlib)

Both have been changed! Our public repository is hyperlinked at the end of our results section, along with a matplotlib acknowledgment and citation. Thanks for point that out!

Micky774 committed:
Updated author details

Micky774 committed:
Updated author details

Micky774 committed:
Updated institutional affiliations

deniederhut commented:
@Micky774 it looks like some of these references are missing DOIs. Could I ask you to add those in?

Micky774 committed:
Trimmed paper to fit 8 pages

Micky774 committed:
Updated bib

Micky774 commented:
Updated to add DOIs to those that have them

mepa commented on papers/zain/mybib.bib:
```diff
@@ -0,0 +1,456 @@
+	file = {Quinn et al. - 2011 - Novel use of differential image velocity invariant.pdf:/home/sonia/Zotero/storage/37EWYT9F/Quinn et al. - 2011 - Novel use of differential image velocity invariant.pdf:application/pdf}
+}
+
+@misc{Cho2014LearningPR,
```
Please add DOI 10.3115/v1/D14-1179

mepa commented on papers/zain/mybib.bib:
```diff
@@ -0,0 +1,456 @@
+	file = {Jégou et al. - 2017 - The One Hundred Layers Tiramisu Fully Convolution.pdf:/home/sonia/Zotero/storage/PNI9PT89/Jégou et al. - 2017 - The One Hundred Layers Tiramisu Fully Convolution.pdf:application/pdf}
+}
+
+@conference{fu_extracting_2004,
```
Please add DOI https://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.185.1179

mepa commented on papers/zain/mybib.bib:
```diff
@@ -0,0 +1,456 @@
+    archivePrefix={arXiv},
+    primaryClass={cs.CL}
+}
+@INPROCEEDINGS{LSTM1999,
```
https://doi.org/10.1162/089976600300015015

mepa reviewed (changes requested):
Hi @Micky774, I'm a conference proceedings co-chair doing a final review of formatting. I noticed a few specific DOIs missing. Could you please add those, plus check for any other references where DOIs are missing and add them in?

Micky774 committed:
Added missing DOIs

Micky774 commented:
Updated, thank you for pointing those out!

Micky774 commented:
As far as I know, some sources don't have DOIs, or at least I couldn't find any for them. Specifically some arXiv preprints and software citations such as Pytorch. I have very little experience with this so please let me know if they indeed also need/have DOIs. Thank you very much for your help.

mepa commented on papers/zain/mybib.bib:
```diff
@@ -0,0 +1,459 @@
+  biburl    = {https://dblp.org/rec/journals/corr/abs-1803-07534.bib},
+  bibsource = {dblp computer science bibliography, https://dblp.org}
+}
+@article{huang_densely_2018,
```
DOI: 10.1109/CVPR.2017.243

mepa commented on papers/zain/mybib.bib:
```diff
@@ -0,0 +1,459 @@
+	file = {Farnebäck - 2003 - Two-Frame Motion Estimation Based on Polynomial Ex.pdf:/home/sonia/Zotero/storage/BA6IB2MF/Farnebäck - 2003 - Two-Frame Motion Estimation Based on Polynomial Ex.pdf:application/pdf}
+}
+
+@article{resnet,
```
DOI: 10.1109/CVPR.2016.90

mepa commented on papers/zain/mybib.bib:
```diff
@@ -0,0 +1,459 @@
+	file = {Beauchemin and Barron - 1995 - The computation of optical flow.pdf:/home/sonia/Zotero/storage/MQE7DUV8/Beauchemin and Barron - 1995 - The computation of optical flow.pdf:application/pdf}
+}
+
+@article{jegou_one_2017,
```
DOI: 10.1109/CVPRW.2017.156

Micky774 committed:
Added DOIs

mepa commented on papers/zain/mybib.bib:
```diff
@@ -0,0 +1,459 @@
+    primaryClass={stat.ML}
+}
+
+@article{introVAE,
```
DOI: 10.1561/2200000056

mepa commented on papers/zain/mybib.bib:
```diff
@@ -0,0 +1,459 @@
+}
+
+
+@article{Numpy2011,
```
DOI: 10.1109/MCSE.2011.37

mepa commented:
Hi, @Micky774. Thanks for updating  so quickly.

I also did not find a DOI for the PyTorch paper, but found a few more for some of the arXiv papers. I did this by copying/pasting the title to a web search and looking at the top few hits. @deniederhut may know if there's a better way?

In any case, after those changes it looks like this paper is good to go. Great job!

Micky774 committed:
Added more DOIs

Micky774 commented:
@mepa Added in the new set of DOIs. Gotcha, I'll try to be more careful about looking through those sources and finding their associated DOIs. I appreciate all your help, and thank you very much!
